### PR TITLE
Piper/limit connections to single ip and dial in and out ratio

### PR DESCRIPTION
**Pull request by pipermerriam**
Replaces #920

### CODE REVIEW

**Comment by pipermerriam** [comment]
@williambannas FYI, I decided to go ahead and do the final cleanup on this myself.  Thanks for getting the majority of the work done.

**Comment by pipermerriam** [comment]
@williambannas we'd love to have you continue contributing.  I'd recommend taking some time to learn your way around `git` and the github workflow.  The rebasing article I linked you might be a good place to start.

**Comment by pipermerriam** [comment]
@carver sanity check review requested since I made some additional changes.

File: p2p/peer.py [hunk 1]
```diff
@@ -585,6 +585,15 @@ def __len__(self):
     def is_full(self) -> bool:
         return len(self) >= self.max_peers
 
+    def is_valid_connection_candidate(self, candidate: Node) -> bool:
+        # connect to no more then 2 nodes with the same IP
+        nodes_by_ip = groupby(
+            operator.attrgetter('remote.address.ip'),
```
Comment (cburgdorf) [inline p2p/peer.py]: Just learned about the `operator` module. Pretty cool to have all operators available as functions :snake:

File: p2p/server.py [hunk 2]
```diff
@@ -264,6 +267,19 @@ def _make_peer_pool(self) -> PeerPool:
...
+        total_peers = len(self.peer_pool.connected_nodes)
+        inbound_peer_count = len([
+            peer
+            for peer
+            in self.peer_pool.connected_nodes.values()
+            if peer.inbound
+        ])
+        if total_peers > 1 and inbound_peer_count / total_peers > DIAL_IN_OUT_RATIO:
+            # make sure to have at least 1/4 outbound connections
+            peer.disconnect(DisconnectReason.useless_peer)
```
Comment (cburgdorf) [inline p2p/server.py]: So, `DisconnectReason.useless_peer` is really the right reason for this case?

File: tests/p2p/test_server.py [hunk 3]
```diff
@@ -44,8 +44,15 @@ def get_open_port():
 INITIATOR_REMOTE = Node(INITIATOR_PUBKEY, INITIATOR_ADDRESS)
 
 
-class MockPeerPool:
+class MockPeerPool():
     is_full = False
+    connected_nodes = {}
```
Comment (cburgdorf) [inline tests/p2p/test_server.py]: I think we should have test for this or maybe at least open an issue to not forget about them.

File: p2p/server.py (hunk 2 above)
Comment (pipermerriam) [reply p2p/server.py]: No, my mistake.  updated to use `too_many_peers` (couldn't find what go-ethereum does)

File: tests/p2p/test_server.py [hunk 4]
```diff
@@ -44,8 +44,15 @@ def get_open_port():
 INITIATOR_REMOTE = Node(INITIATOR_PUBKEY, INITIATOR_ADDRESS)
 
 
-class MockPeerPool:
+class MockPeerPool():
```
Comment (carver) [inline tests/p2p/test_server.py]: Added parens unnecessarily?

**Pull request merged**